scene/main/viewport.cpp

```diff
@@ -2607,6 +2633,7 @@ void Viewport::_drop_mouse_focus() {
 void Viewport::_drop_physics_mouseover(bool p_paused_only) {
 	physics_has_last_mousepos = false;
 
+	// TODO: deduplicate this with the erasing in _process_picking (send to 2D).
```

Comment:
Thanks for the comment. Unfortunately, it's not as simple as you suggest. I refactored the code while keeping the logic as is; do you know if I can further simplify it? ie. do you know why the shapes are removed only when they are castable to CollisionObject2D in one case and in all cases in the other?

On a more general note, is there a place for documentation of private methods where I should document this better than with a comment in .h?

Reply:
After looking at the code, it looks like it should erase from the list outside of the CollisionObject2D check in both cases to me.

The difference was introduced in #45775.
@RandomShaper Do you remember if there was a specific reason for this difference or if it could be a typo?

For more context, the original logic when triggering exit events is:
https://github.com/godotengine/godot/blob/277db14b647d3184b341a359020c0acc4549afaf/scene/main/viewport.cpp#L735-L742

But in `_drop_physics_mouseover`, `to_erase.push_back(E)` is done only after casting to `CollisionObject2D`:
https://github.com/godotengine/godot/blob/277db14b647d3184b341a359020c0acc4549afaf/scene/main/viewport.cpp#L2604-L2614

If we can confirm they both should be like the first case, it will be possible to further simplify the code, and it would need a fix on 3.x branch as well.

@sygi About documenting private methods, there's no specific place for that. You can leave extra comments in the cpp, either right before or within the method if you want to add more details.

Reply:
The second case is different because because of `p_paused_only`. If it's `true`, the node shall not be removed.

That said, I think it may be good to erase it as well if the check for a valid instance fails (the `if (o)` check).

Furthermore, refactoring the code so if `p_paused_only` is checked in an outer scope may help, by allowing to have a loop similar to the one in the first snippet (unconditional removal). However, an additional variant (that checks `co->can_process()`) will still be needed, so I'm not sure if that could help with what is being done in this PR.

Reply:
Could you clarify if the logic in #528d029 is correct or not? (it's keeps `p_paused_only`)

Reply:
I think it isn't.

`_drop_physics_mouseover()` is called with `p_paused_only = true` every frame, to send `_mouse_exit` to nodes that have just gotten paused. So, if `to_erase.push_back(E);` happens regardless `p_paused_only && co->can_process()`, any node that gets hovered will immediately be removed from being considered hovered, so it won't get `_mouse_exit` when unhovering actually happens.

In other words, a node should be "unmarked" from being hovered (`to_erase.push_back(E);`) if, and only if, it's as well notified to have been unhovered (`co->_mouse_exit()`).

Reply:
Thanks for the comment. Sorry about that, but I'm still a bit confused. You say:
> it may be good to erase it as well if the check for a valid instance fails (the if (o) check)

In the code above I only moved the `to_erase` check to happen if `!o || !co`. The handling of the case when `p_paused_only && co->can_process()` is the same (no erasure). In particular, under the assumption that `co != null`, `co->_move_exit()` happens iff `to_erase.push_back(E)`.

Are we speaking about splitting handling of `p_paused_only` and `co->can_process`? should we erase when `p_paused_only` even if `!co`?

Reply:
Yeah, I think it's fine then, `continue` will prevent removal from the list when not processed yet. Just waiting for confirmation from @RandomShaper, but the current state of the PR should be good to go.

Reply:
Today, with a pair of fresh eyes, I've been able to read the code correctly. It's fine indeed.